### ISSUE TITLE
Failed assertion during two-finger scrolling in a11y mode

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#11672]
Sometimes I see this exception when scrolling on Android with two fingers in a11y mode:
```
E/flutter (10226): [ERROR:lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (10226): 'package:flutter/src/gestures/converter.dart': Failed assertion: line 131 pos 18: '!state.down': is not true.
E/flutter (10226): #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:35)
E/flutter (10226): #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:31)
E/flutter (10226): #2      PointerEventConverter.expand (package:flutter/src/gestures/converter.dart:131:18)
E/flutter (10226): #3      _SyncIterator.moveNext (dart:core-patch/core_patch.dart:61)
E/flutter (10226): #4      ListQueue.addAll (dart:collection/queue.dart:635)
E/flutter (10226): #5      BindingBase&SchedulerBinding&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:46:27)
E/flutter (10226): #6      _invoke1 (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:100)
E/flutter (10226): #7      _dispatchPointerDataPacket (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:58)
```

**Comment by goderbauer** [comment]
New development: The app now becomes unresponsive for a few seconds after this assert fires. Increasing priority.

**Comment by goderbauer** [comment]
Android is actually dispatching two down events for the same pointer/finger in a row, which is throwing off our state machine that keeps track of the status of all pointers:

```
MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=1, x[0]=694.0, y[0]=1755.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823827, downTime=66823827, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=1, x[0]=694.0, y[0]=1755.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=1, eventTime=66823848, downTime=66823827, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=1, x[0]=694.0, y[0]=1755.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=1, eventTime=66823868, downTime=66823827, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=1, x[0]=694.0, y[0]=1754.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823878, downTime=66823827, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_UP, actionButton=0, id[0]=1, x[0]=693.0, y[0]=1754.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823888, downTime=66823827, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=725.0, y[0]=1546.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823888, downTime=66823898, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_CANCEL, actionButton=0, id[0]=0, x[0]=725.0, y[0]=1546.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823903, downTime=66823898, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=725.0, y[0]=1543.0, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=66823898, downTime=66823898, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_POINTER_DOWN(1), actionButton=0, id[0]=0, x[0]=725.0, y[0]=1546.0, toolType[0]=TOOL_TYPE_FINGER, id[1]=1, x[1]=693.0, y[1]=1754.0, toolType[1]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=2, historySize=0, eventTime=66823888, downTime=66823898, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_POINTER_DOWN(1), actionButton=0, id[0]=0, x[0]=725.0, y[0]=1543.0, toolType[0]=TOOL_TYPE_FINGER, id[1]=1, x[1]=691.0, y[1]=1754.0, toolType[1]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=2, historySize=0, eventTime=66823898, downTime=66823898, deviceId=7, source=0x1002 }
MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=0, x[0]=725.0, y[0]=1539.0, toolType[0]=TOOL_TYPE_FINGER, id[1]=1, x[1]=689.0, y[1]=1753.0, toolType[1]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=2, historySize=0, eventTime=66823909, downTime=66823898, deviceId=7, source=0x1002 }
```

**Comment by goderbauer** [comment]
I have confirmed that the duplicated down event is most likely a bug in Android's a11y framework and we will have to work around it. We can either just filter out this one specific duplicated event or we can can make Flutter's event processing more robust to expect the unexpected and be ok with duplicated events. (Android a11y folks recommend the latter as they apparently have seen a lot of weirdness around events).

Anyhow, I am moving this below the critical cutoff for M3 as the assert doesn't cause any practical problems currently.

**Comment by Hixie** [comment]
If we implement a workaround, we should do it in the engine's Android code, so that we don't paper over bugs elsewhere.

But the right solution is for the Android a11y framework to fix its bug.

**Comment by goderbauer** [comment]
This will presumably be fixed by the fix we have in mind for https://github.com/flutter/flutter/issues/20517.

**Comment by zoeyfan** [comment]
Given #20517 is fixed now, I assume we can close this one safely, right?
